y/zstd.go

```diff
@@ -19,18 +17,38 @@
 package y
 
 import (
-	"github.com/DataDog/zstd"
+	"sync"
+
+	"github.com/klauspost/compress/zstd"
 )
 
-// CgoEnabled is used to check if CGO is enabled while building badger.
-const CgoEnabled = true
+var (
+	zstdDec *zstd.Decoder
+	zstdEnc *zstd.Encoder
+
+	zstdEncOnce, zstdDecOnce sync.Once
+)
 
 // ZSTDDecompress decompresses a block using ZSTD algorithm.
 func ZSTDDecompress(dst, src []byte) ([]byte, error) {
-	return zstd.Decompress(dst, src)
+	var err error
+	zstdDecOnce.Do(func() {
+		zstdDec, err = zstd.NewReader(nil)
+		AssertTrue(err == nil)
+	})
+	return zstdDec.DecodeAll(src, dst)
 }
 
 // ZSTDCompress compresses a block using ZSTD algorithm.
-func ZSTDCompress(dst, src []byte, compressionLevel int) ([]byte, error) {
-	return zstd.CompressLevel(dst, src, compressionLevel)
+func ZSTDCompress(dst, src []byte, level int) ([]byte, error) {
+	var err error
+	zstdEncOnce.Do(func() {
+		zstdEnc, err = zstd.NewWriter(
+			nil, zstd.WithZeroFrames(true),
+			zstd.WithEncoderLevel(zstd.EncoderLevelFromZstd(level)),
+			zstd.WithEncoderCRC(false))
+
+		AssertTrue(err == nil)
+	})
+	return zstdEnc.EncodeAll(src, dst), nil
```

Comment:
```suggestion
	return zstdEnc.EncodeAll(src, dst[:0]), nil
```

This is likely the problem. Output is appended to make it easier to use in serializers, etc.